text/0035-schema-applies-to-unspecified.md

```diff
@@ -0,0 +1,100 @@
+2. Omitted `appliesTo` means that the action cannot be used in a request. It is used exclusively as an action group.
+3. Disallow empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. Disallow an empty `appliesTo` attribute thus requiring at least one of `principalTypes`, `resourceTypes` or `context` to be specified if `appliesTo` is specified.
+
+This feature request is consistent with the semantics for `appliesTo` proposed in [RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) for the alternate Cedar schema syntax.
+
+## Detailed design
+
+This RFC will require the following changes:
+
+1. The permissive validator must be updated to treat an omitted `appliesTo.principalTypes` / `appliesTo.resourceTypes` as the _unspecified_ entity type, rather than _any_ entity type. The difference is that the _unspecified_ entity type is a particular type that is different from any defined type, whereas _any_ entity type could be the unspecified type or any other type defined in the schema. Note that this change is only required in the _permissive_ validator, because the corresponding change was already made to strict validation as part of [RFC 19](https://github.com/cedar-policy/rfcs/blob/main/text/0019-stricter-validation.md).
+2. The validator must be updated to treat an omitted `appliesTo` attribute as indication that an action does not apply to any principal or resource types.
+3. The schema parser must be updated to reject empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. The schema parser must be updated to reject empty `appliesTo` attributes.
+
+### Notes on Backwards Compatibility
+
+This RFC proposes a **breaking change** that requires a major version bump to Cedar.
+For the sake of argument, say that the current version of Cedar X and that this RFC will be implemented in Cedar Y (which has no other breaking changes).
+The rest of this section is written for users who are upgrading from X to Y.
+
+The changes proposed in this RFC will _only_ affect you if you have one or more of the following in your schema:
+
+1. You intend for an action to apply to _no_ principals or resources, i.e.,
+
+- `appliesTo.principalTypes: []`
+- `appliesTo.resourceTypes: []`
+
+2. You intend for an action to apply to _any_ principal and/or resource types, i.e.,
+
+- omitted `appliesTo` field
+- omitted `appliesTo.principalTypes` field
+- omitted `appliesTo.resourceTypes` field
+
+In **case 1**, in preparation to upgrade, you will need to update your schema to omit the `appliesTo` field.
+Your current schema will result in error when you upgrade to Y.
+If you use your modified schema with X, you will be able to validate policies that will not validate in Y.
+There is _no way_ to write a schema so that validation behavior will be the same between X and Y.
```

Comment:
It seems undesirable that there is no way to write a schema with the same meaning before & after this change. An easy fix would be to leave `principalTypes: []` and `resourceTypes: []` as the only way to specify that an action can apply to no entity types. This is arguably more intuitive too: rather than omitting a field (which otherwise means that something is unspecified) users have to explicitly say that an action should not be used in requests.

Reply:
More concretely, I would propose updating this RFC to the following:
| Scenario | Status Quo | Proposal |
| --------- | -------- | ------- |
| Specified `appliesTo` but omitted `principalTypes`/`resourceTypes` | Authz request can have a principal/resource entity of any type, or the unspecified entity | Authz request with this action can only have unspecified entity as principal/resource |
| `appliesTo.principalTypes: []` AND `appliesTo.resourceTypes: []` | Action cannot be used in authz request | Action cannot be used in authz request  |
| One of `appliesTo.principalTypes: []` OR `appliesTo.resourceTypes: []` | Action cannot be used in authz request | ❌ Not allowed  |
| Omitted `appliesTo` | Authz request can have a principal/resource entity of any type, or the unspecified entity | ❌ Not allowed|

In all cases, if `context` is not specified in the `appliesTo` field, then it defaults to the empty record `{}`.

Reply:
An alternative that also makes sense to me: combine the first and last rows of the table above, so that omitting `appliesTo` has the same meaning as omitting `principalTypes` and `resourceTypes`.

Reply:
For case 1: I'm confused. If you wrote `principalTypes: []` or `resourceTypes: []` before, in order to indicate that an action cannot be used in requests, then you should change your schema to remove the `appliesTo` clause instead. So you can still express in the proposal what you could express before. 

Moreover, it's much more obvious what you intend: With no `appliesTo` it's clear that this action is not "applyable" in a request. Part of the motivation for this proposal is to rule out nonsensical and/or confusing specs. @max2me specifically didn't like the idea that once you had `principalTypes: []` then it didn't matter what else you specified for the other bits -- you can't use this action in requests. The proposed change eliminates this confusion. I much prefer its clarity.

This change will be made evident by the validator. If you previously wrote `principalTypes: []` etc. then under this proposal you'll get an error. The fix is clear: remove the `appliesTo`. If you previously omitted `appliesTo` entirely but used the action in policies, then you'll get an error under this proposal, which should prompt you to add `appliesTo { }` instead.

For case 2: as I wrote in a comment above, I don't think we actually supported the idea that you could apply to _any_ type, in the sense that we specifically considered every type during validation.

Reply:
> If you wrote `principalTypes: []` or `resourceTypes: []` before, in order to indicate that an action cannot be used in requests, then you should change your schema to remove the `appliesTo` clause instead. So you can still express in the proposal what you could express before.

Yes, both proposals (the original RFC and my table above) allow you to express the same thing. The difference is that one is backwards compatible: it means the same thing in both before and after the proposed change. With the proposal in the original RFC, there is no way to update your schema to prepare for this change -- you have to update your schema simultaneously with the upgrade to the new version of Cedar.

> once you had `principalTypes: []` then it didn't matter what else you specified for the other bits

I also think that allowing only one of `principalTypes` / `resourceTypes` to be an empty list is confusing (since it means the same thing as having both be the empty list). I propose disallowing this above.

Reply:
Late to the conversation, but I prefer the original proposal given in the RFC.  Omitting `appliesTo` is clearer (to my eyes) and more ergonomic than having to set _both_ `principalTypes` and `resourceTypes` to the empty list. 

Since the change will be breaking no matter what, I'd vote for adopting the simplest long-term solution, even though it incurs more short-term inconvenience (having to update the schema simultaneously with the upgrade to the new version of Cedar).

Reply:
One key thing to understand is that these updates are not atomic. People will have to update their policies and schemas to be forwards compatible, wait some non-zero time period (potentially weeks or months), then perform the update to cedar 3.0. 

In that interim time (during those few weeks/months) there is no path for customers to express their actual intent in the schema. This is because the same syntax means different things in both proposals.

Reply:
Marking this thread as resolved since the majority opinion seems to be:
> Omitting appliesTo is clearer (to my eyes) and more ergonomic than having to set both principalTypes and resourceTypes to the empty list.

Reply:
What is the resolution to the issue raised in this thread about

> It seems undesirable that there is no way to write a schema with the same meaning before & after this change

> With the proposal in the original RFC, there is no way to update your schema to prepare for this change -- you have to update your schema simultaneously with the upgrade to the new version of Cedar.

These seemed to be the motivating arguments for the other side, even though I agree that the majority feel that the RFC's current proposal is clearer and more ergonomic than having to set both `principalTypes` and `resourceTypes` to the empty list.

Reply:
Good points Craig. I've updated the RFC text to explicitly list the proposed alternative, and started a fresh discussion thread: https://github.com/cedar-policy/rfcs/pull/35#discussion_r1386794236

Resolving this thread to keep the discussion organized.